### DISCUSSION TITLE
Options for improving precision

### DISCUSSION THREAD

**Discussion by mattjohnsonpint** [mattjohnsonpint/GeoTimeZone#56] [Polls]
In #10 we talked about ways to improve precision algorithmically.  However, one of the easiest ways we could improve precision is to build the data at a higher geohash precision.

Reference: 
- https://en.wikipedia.org/wiki/Geohash#Digits_and_precision_in_km
- https://docs.quadrant.io/quadrant-geohash-algorithm

For this library, the practical options would be:

| Level | Grid Area | Error | Data Size |
|---|---|---|---|
| 5 | ≤ 4.89km X 4.89km | ±2.4 km | 3 MB |
| 6 | ≤ 1.22km X 0.61km | ±0.61 km | 18 MB |
| 7 | ≤ 153m X 153m | ±76 m | 108 MB |

Currently we're using level 5.

**Comment by jnyrup** [comment]
I really like the idea of having multiple precisions to pick from, as the precision needed is context specific, both in terms of domain and available machine resources. 

I guess having multiple nuget packages could create a diamond dependency problem?

Having the data in separate files also let's you dynamically pick one.

**Reply by jnyrup** [reply]
Another (minor) matter with respect to file size is that as the data is embedded in the dll, i.e. once per TFM, the nuget packages grows by a factor of four.
So increasing the data from 3mb to 18mb, would approximately increase the size of the nuget package by 60 mb (minus some amount of compression).

**Reply by mattjohnsonpint** [reply]
Sure, but does the package size really matter? The resulting compiled project will only use the ones it targets.

**Reply by jnyrup** [reply]
It only affects restore time when the package is not in any available nuget cache. And even there it might be insignificant for 60mb. 

But a 4*108 mb large nuget package seems very large, at least to me.

**Reply by mattjohnsonpint** [reply]
True.  But that gives me another idea... https://github.com/mattjohnsonpint/GeoTimeZone/discussions/56#discussioncomment-4824779

**Comment by benberlin7** [comment]
I think there should be an easy-to-go setup which does not demand users to think about the geohash precision issue in the first place, so level 5 in the standard procedure is completely fine IMHO. However, it'd be nice to enable experienced users to optionally upgrade the precision of the data if needed.

**Comment by mattjohnsonpint** [comment]
Another approach I just thought of, would be to put all three data files into a separate folder within the NuGet package.  Then a custom MSBuild target could be included in the package that would embed one of the three files into the target assembly at build time.  One would control the precision in their csproj, for example:

```xml
<PropertyGroup>
    <GeoTimeZonePrecision>7</GeoTimeZonePrecision>
</PropertyGroup>
```

**Reply by mattjohnsonpint** [reply]
Alternatively, it could just copy the file into the output directory.  It wouldn't need to be an embedded resource.

**Reply by benberlin7** [reply]
I think a huge size of the repo would be a blocker for us.

**Reply by mattjohnsonpint** [reply]
How about if the chosen data file was downloaded from GitHub at build time then?

**Reply by jnyrup** [reply]
I see some scenarios, where delaying the downloading to build time is not desirable.
E.g. if the files are not cached somewhere, you cannot build if you're offline or if GitHub has an outage.

**Reply by AdrianoAE** [reply]
Is there any update to this? Having the option to go with the highest precision would be very useful

**Reply by mattjohnsonpint** [reply]
Sorry, I haven't had time to work on this yet.

**Reply by jnyrup** [reply]
Anything the community can help with Matt, or do you prefer to play around with it yourself when you find the time?

**Reply by rajesh-patel-sh** [reply]
Seems like a great solution for this problem!

**Comment by Zooce** [comment]
Out of curiosity (and I'd like to test it out), what changes would need to be made to increase the precision to 6 digits? I see a `Precision` property in Geohash.cs and GeohashTree.cs. Is it really as simple as changing those two properties to 6?

**Reply by mattjohnsonpint** [reply]
I believe so. Change that and rebuild the data.

(Sorry, it's been quite a long time since I last looked at this code.)

**Reply by Zooce** [reply]
Almost worked - also had to increase the `LineLength` property of TimeZoneFileReader to 9 so that `GetLine` would always produce 3 digits.
